codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -285,193 +314,498 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
-            }
-        }
-        return future;
-    }
+    private static final class DefaultHttp2StreamChannel extends AbstractChannel implements Http2StreamChannel {
 
-    @SuppressWarnings("unchecked")
-    private static void initOpts(Channel channel, Map<ChannelOption<?>, Object> opts) {
-        if (opts != null) {
-            for (Entry<ChannelOption<?>, Object> e: opts.entrySet()) {
-                try {
-                    if (!channel.config().setOption((ChannelOption<Object>) e.getKey(), e.getValue())) {
-                        logger.warn("Unknown channel option: " + e);
-                    }
-                } catch (Throwable t) {
-                    logger.warn("Failed to set a channel option: " + channel, t);
-                }
-            }
-        }
-    }
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<DefaultHttp2StreamChannel> OUTBOUND_FLOW_CONTROL_WINDOW_UPDATER;
 
-    @SuppressWarnings("unchecked")
-    private static void initAttrs(Channel channel, Map<AttributeKey<?>, Object> attrs) {
-        if (attrs != null) {
-            for (Entry<AttributeKey<?>, Object> e: attrs.entrySet()) {
-                channel.attr((AttributeKey<Object>) e.getKey()).set(e.getValue());
-            }
-        }
-    }
+        /**
+         * Used by subclasses to queue a close channel within the read queue. When read, it will close
+         * the channel (using Unsafe) instead of notifying handlers of the message with {@code
+         * channelRead()}. Additional inbound messages must not arrive after this one.
+         */
+        private static final Object CLOSE_MESSAGE = new Object();
+        /**
+         * Used to add a message to the {@link ChannelOutboundBuffer}, so as to have it re-evaluate its writability
+         * state.
+         */
+        private static final Object REEVALUATE_WRITABILITY_MESSAGE = new Object();
+        private static final ChannelMetadata METADATA = new ChannelMetadata(false, 16);
+        private static final ClosedChannelException CLOSED_CHANNEL_EXCEPTION = ThrowableUtil.unknownStackTrace(
+                new ClosedChannelException(), DefaultHttp2StreamChannel.class, "doWrite(...)");
+        /**
+         * Number of bytes to consider non-payload messages, to determine when to stop reading. 9 is
+         * arbitrary, but also the minimum size of an HTTP/2 frame. Primarily is non-zero.
+         */
+        private static final int ARBITRARY_MESSAGE_SIZE = 9;
+
+        private final ChannelHandlerContext ctx;
+        private final ChannelFutureListener firstFrameWriteListener;
+        private final Http2StreamChannelConfig config = new Http2StreamChannelConfig(this);
+        private final Queue<Object> inboundBuffer = new ArrayDeque<Object>(4);
+        private final Http2FrameStream stream;
+
+        private boolean closed;
+        private boolean readInProgress;
+        private MessageSizeEstimator.Handle sizeEstimatorHandle;
+
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)